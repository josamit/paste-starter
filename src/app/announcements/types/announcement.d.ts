import { AlertVariants } from "@twilio-paste/alert";

export type Announcement = {
  id: string;
  content: string;
  valid_from?: string;
  valid_until?: string;
  variant: AlertVariants;
  can_be_dismissed: boolean;
};
