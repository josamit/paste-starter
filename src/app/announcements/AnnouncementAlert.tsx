import React, { FC, useMemo } from "react";
import { Alert, Anchor, Box, Text } from "@twilio-paste/core";
import { AlertVariants } from "@twilio-paste/alert";
import { marked } from "marked";

export type AnnouncementProps = {
  id: string;
  content: string;
  isVisible: boolean;
  variant: AlertVariants;
  onDismiss?: () => void;
};

export const markdownToReactNode = (content: string, id: string) => {
  const result = marked.lexer(content);
  const reactTokens: React.ReactElement[] = [];

  result.forEach((topLevelToken) => {
    if ("tokens" in topLevelToken) {
      const tokens = topLevelToken.tokens || [];
      tokens.forEach((token) => {
        switch (token.type) {
          case "text":
            reactTokens.push(
              <Text as="span" key={token.text} id={id}>
                {token.text}
              </Text>
            );
            break;
          case "link":
            // Not using the default parsed variant to show that we can put whatever element we want
            const shouldShowExternal = token.href.includes("twilio.com");
            reactTokens.push(
              <Anchor
                href={token.href}
                showExternal={shouldShowExternal}
                target={shouldShowExternal ? "_blank" : "_self"}
                key={token.text}
                id={id}
              >
                {token.text}
              </Anchor>
            );
            break;
          case "strong":
            reactTokens.push(
              <Text
                as="span"
                fontWeight="fontWeightBold"
                key={token.text}
                id={id}
              >
                {token.text}
              </Text>
            );
            break;
          case "em":
            reactTokens.push(
              <Text
                as="span"
                fontFamily="fontFamilyCode"
                color="colorTextBrandHighlight"
                key={token.text}
                id={id}
              >
                {token.text}
              </Text>
            );
            break;
          default:
            reactTokens.push(
              <span key={token.raw} id={id}>
                {token.raw}
              </span>
            );
            break;
        }
      });
    }
  });

  return <div>{reactTokens}</div>;
};

const AnnouncementAlert: FC<AnnouncementProps> = ({
  isVisible,
  onDismiss,
  variant,
  content,
  id,
}) => {
  const memoizedNodes = useMemo(
    () => markdownToReactNode(content, id),
    [content, id]
  );

  if (isVisible) {
    return (
      <Box id={id} marginBottom="space70" data-testid={id}>
        <Alert variant={variant} onDismiss={onDismiss} data-testid={id}>
          {memoizedNodes}
        </Alert>
      </Box>
    );
  }

  return <></>;
};

export default AnnouncementAlert;
