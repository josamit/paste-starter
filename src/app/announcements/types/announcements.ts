import { Announcement } from "@app/announcements/types/announcement";

export const announcements: Announcement[] = [
  {
    id: "first_announcement",
    content: "This is a sample announcment. You can dissmiss it safely.",
    valid_from: "2021-01-01T00:00:00Z",
    valid_until: "2025-12-31T23:59:59Z",
    variant: "neutral",
    can_be_dismissed: true,
  },
];
