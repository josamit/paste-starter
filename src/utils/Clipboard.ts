import { UseToasterReturnedProps } from "@twilio-paste/core";

const handleErr = (
  toaster: UseToasterReturnedProps,
  itemName: string,
  err: unknown
) => {
  toaster.push({
    message: `Oops, unable to copy ${itemName} to clipboard. ${err}`,
    variant: "error",
  });
};

const handleSuccess = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "success",
  });
};

const handleWarning = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "warning",
  });
};

export function copyToClipboardSync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  const textArea = document.createElement("textarea");
  textArea.value = textToCopy;

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();

  try {
    const copyResult = document.execCommand("copy");
    const message = copyResult
      ? `${itemName} was successfully copied to your clipboard.`
      : `${itemName} could not be copied to your clipboard. Please try again.`;
    if (copyResult) {
      handleSuccess(toaster, message);
    } else {
      handleWarning(toaster, message);
    }
  } catch (err) {
    handleErr(toaster, itemName, err);
  }

  document.body.removeChild(textArea);
}

export function copyToClipboardAsync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  if (!navigator.clipboard) {
    copyToClipboardSync(textToCopy, itemName, toaster);
    return;
  }

  navigator.clipboard
    .writeText(textToCopy)
    .then(() =>
      handleSuccess(
        toaster,
        `${itemName} was successfully copied to your clipboard.`
      )
    )
    .catch((err) => handleErr(toaster, itemName, err));
}
