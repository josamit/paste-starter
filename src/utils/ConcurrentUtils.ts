/*
 * takes an array of items and a function that accepts an item at a time and returns a promise
 * returns a promise that resolves to an array of results
 */
export function mapConcurrent<T>(
  items: string | any[],
  maxConcurrent: number,
  fn: (arg0: T, arg1: number) => Promise<any>
) {
  let index = 0;
  let currentlyInFlightRequests = 0;
  let completedRequests = 0;
  const results = new Array(items.length);
  let stopExecution = false;

  return new Promise(function (resolve, reject) {
    function run() {
      // launch as many as we're allowed to
      while (
        !stopExecution &&
        currentlyInFlightRequests < maxConcurrent &&
        index < items.length
      ) {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        runNext();
      }
      // if all are done, then resolve parent promise with results
      if (completedRequests === items.length) {
        resolve(results);
      }
    }

    function runNext() {
      const i = index;
      // eslint-disable-next-line no-plusplus
      ++currentlyInFlightRequests;
      // eslint-disable-next-line no-plusplus
      fn(items[index], index++).then(
        function (val) {
          // eslint-disable-next-line no-plusplus
          ++completedRequests;
          // eslint-disable-next-line no-plusplus
          --currentlyInFlightRequests;
          results[i] = val;
          run();
        },
        function (err) {
          // set flag so we don't launch any more requests
          stopExecution = true;
          reject(err);
        }
      );
    }

    run();
  });
}
