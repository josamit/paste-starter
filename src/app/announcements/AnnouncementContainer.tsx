import React, { FC } from "react";
import {
  announcementsNamespacedLSKey,
  DISMISSED,
  useAnnouncement,
} from "@app/announcements/hooks/useAnnouncement";
import { Announcement } from "./types/announcement";
import AnnouncementAlert from "./AnnouncementAlert";

export type AnnouncementProps = Announcement;

const AnnouncementContainer: FC<AnnouncementProps> = ({
  valid_from,
  valid_until,
  variant,
  id,
  content,
  can_be_dismissed,
}) => {
  const [isVisible, setVisible] = useAnnouncement(id, valid_from, valid_until);

  const dismissNotification = () => {
    localStorage.setItem(announcementsNamespacedLSKey(id), DISMISSED);
    setVisible(false);
  };

  return (
    <AnnouncementAlert
      id={id}
      variant={variant}
      content={content}
      isVisible={isVisible}
      onDismiss={can_be_dismissed ? dismissNotification : undefined}
    />
  );
};

export default AnnouncementContainer;
