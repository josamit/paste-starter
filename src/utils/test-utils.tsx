import * as React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Theme } from "@twilio-paste/core/theme";
import { MockQueryManagerContext } from "@context/QueryManagerContext";
import { WindowSizeContext } from "@context/WindowSizeProvider";
import { MemoryRouter, Route, Switch } from "react-router-dom";
import { DarkModeContext } from "@context/DarkModeContext";
import ToastsProvider from "@context/ToastsProvider";
import { createGetForm } from "@src/utils/form";
import { Box } from "@twilio-paste/core/dist/box";

export function renderWithFormUtils(...args: Parameters<typeof rtlRender>) {
  const [children, ...restArgs] = args;

  const results = rtlRender(<Box>{children}</Box>, ...restArgs);
  const getForm = createGetForm(results);

  return {
    ...results,
    getForm,
  };
}

const renderComponent = (
  ui: React.ReactElement,
  {
    initialEntry = "/route/1",
    initialRoute = "/route/:appSid",
    queryParams = {
      accountSid: "ACxxx",
      activeTab: "a2pBrandsReconciliation",
    },
    container = undefined,
    mockToggleMode = jest.fn(),
    setQueryParams = jest.fn(),
    ...renderOptions
  } = {}
) => {
  const Wrapper: React.FC = ({ children }) => (
    <MockQueryManagerContext
      value={{
        queryParams,
        setQueryParams,
      }}
    >
      <WindowSizeContext.Provider
        value={{ width: 1400, height: 800, breakpointIndex: 450 }}
      >
        <Theme.Provider theme="default" disableAnimations>
          <DarkModeContext.Provider
            value={{
              theme: "default",
              toggleMode: mockToggleMode,
              componentMounted: true,
            }}
          >
            <ToastsProvider>
              <MemoryRouter initialEntries={[initialEntry]}>
                <Switch>
                  <Route path={initialRoute}>{children}</Route>
                </Switch>
              </MemoryRouter>
            </ToastsProvider>
          </DarkModeContext.Provider>
        </Theme.Provider>
      </WindowSizeContext.Provider>
    </MockQueryManagerContext>
  );

  return renderWithFormUtils(ui, {
    wrapper: Wrapper,
    container,
    ...renderOptions,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderComponent };
