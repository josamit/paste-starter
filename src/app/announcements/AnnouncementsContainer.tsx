import React, { FC } from "react";
import { Box } from "@twilio-paste/core";
import { announcements as defaultAnnouncements } from "@app/announcements/types/announcements";
import { Announcement } from "./types/announcement";
import AnnouncementContainer from "./AnnouncementContainer";

export type AnnouncementsProps = {
  presetAnnouncements?: Announcement[];
};

const AnnouncementsContainer: FC<AnnouncementsProps> = ({
  presetAnnouncements = defaultAnnouncements,
}) => {
  return (
    <>
      {presetAnnouncements.map(
        ({
          id,
          content,
          variant,
          valid_from,
          valid_until,
          can_be_dismissed,
        }) => (
          <AnnouncementContainer
            id={id}
            content={content}
            variant={variant}
            valid_from={valid_from}
            valid_until={valid_until}
            data-tesid={`${id}-container`}
            can_be_dismissed={can_be_dismissed}
          />
        )
      )}
    </>
  );
};

export default AnnouncementsContainer;
