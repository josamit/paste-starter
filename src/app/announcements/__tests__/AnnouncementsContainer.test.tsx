import * as React from "react";
import { renderComponent } from "@src/utils/test-utils";
import AnnouncementsContainer from "@app/announcements/AnnouncementsContainer";
import { announcements } from "@app/announcements/types/announcements";

describe("<AnnouncementsContainer />", () => {
  it("should render correctly", async () => {
    const { container, getByText, getByTestId } = renderComponent(
      <AnnouncementsContainer presetAnnouncements={announcements} />
    );

    getByText("If you are using this page for a compliance audit, please");
    getByText("DO NOT MODIFY OR DELETE");

    expect(
      container.querySelector('a[href="mailto:product-compliance@twilio.com"]')
    ).not.toBeNull();

    expect(container).toMatchSnapshot();
  });
});
