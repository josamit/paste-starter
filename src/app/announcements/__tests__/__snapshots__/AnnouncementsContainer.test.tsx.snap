// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnnouncementsContainer /> should render correctly 1`] = `
<div>
  <div
    class="css-ql39e"
  >
    <div
      class="css-roynbj"
      data-paste-element="BOX"
      data-testid="first_announcement"
      id="first_announcement"
    >
      <div
        class="css-htbnjw"
        data-paste-element="ALERT"
        data-testid="first_announcement"
        role="status"
      >
        <div
          class="css-1sekzvp"
          data-paste-element="MEDIA_OBJECT"
          display="flex"
        >
          <div
            class="css-1llkq7k"
            data-paste-element="MEDIA_FIGURE"
            display="flex"
          >
            <span
              class="css-vha63f"
              color="colorTextIconNeutral"
              data-paste-element="ALERT_ICON"
              display="block"
            >
              <svg
                aria-hidden="false"
                height="100%"
                role="img"
                viewBox="0 0 20 20"
                width="100%"
              >
                <title>
                  information: 
                </title>
                <path
                  d="M10 2a8 8 0 110 16 8 8 0 010-16zm0 7c-.26.01-.528.08-.707.24-.178.158-.292.355-.292.575v3.37c0 .21.11.428.292.577.173.14.457.247.707.237.26-.008.528-.078.707-.237.178-.16.292-.355.292-.577v-3.37a.766.766 0 00-.292-.575A1.14 1.14 0 0010 9zm0-3l-.266.036c-.17.047-.318.131-.441.253l-.156.198A.958.958 0 009 6.98v.04a.728.728 0 00.086.378c.04.121.108.227.206.315.09.094.197.163.321.202.12.06.248.09.386.084l.266-.035c.17-.047.318-.131.441-.253l.156-.198A.958.958 0 0011 7.02v-.04a.728.728 0 00-.086-.378.721.721 0 00-.206-.315.757.757 0 00-.321-.202.767.767 0 00-.386-.084z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </div>
          <div
            class="css-1i39mnm"
            data-paste-element="MEDIA_BODY"
          >
            <div>
              <span
                class="css-1n4qplb"
                color="colorText"
                data-paste-element="TEXT"
                font-size="fontSize30"
                id="first_announcement"
              >
                This is a sample announcement. You can dismiss it safely.
              </span>
            </div>
          </div>
          <span
            class="css-17yrt8g"
            data-paste-element="MEDIA_FIGURE"
            display="flex"
          >
            <button
              aria-busy="false"
              class="css-140t9ab"
              color="colorTextIcon"
              cursor="pointer"
              data-paste-element="ALERT_DISMISS_BUTTON"
              display="inline-block"
              font-family="fontFamilyText"
              font-weight="fontWeightSemibold"
              text-decoration="none"
              type="button"
              width="auto"
            >
              <span
                aria-hidden="false"
                class="css-cfzbzn"
                data-paste-element="BOX"
                display="flex"
                opacity="1"
                text-decoration="inherit"
              >
                <span
                  class="css-1kzjb37"
                  color="colorTextIcon"
                  data-paste-element="ALERT_DISMISS_ICON"
                  display="block"
                >
                  <svg
                    aria-hidden="false"
                    height="100%"
                    role="img"
                    viewBox="0 0 20 20"
                    width="100%"
                  >
                    <title>
                      Dismiss alert
                    </title>
                    <path
                      d="M15.16 13.514L11.645 10l3.515-3.514a1.165 1.165 0 00-1.646-1.646L10 8.355 6.486 4.84A1.165 1.165 0 004.84 6.486L8.355 10 4.84 13.514a1.165 1.165 0 001.646 1.646L10 11.645l3.514 3.515a1.165 1.165 0 001.646-1.646z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
